File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Integer[]
     */
    function preorderTraversal($root) {
        
        if($root->val === null){
            return [];
        }
        
        return array_merge(
            [
                $root->val
            ], 
            $this->preorderTraversal($root->left), 
            $this->preorderTraversal($root->right)
        );
    }
}

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Result struct {
    Ans []int
}

func preorderTraversal(root *TreeNode) []int {
    res := Result{}
    res.dfs(root)
    return res.Ans
}

func (r *Result) dfs(node *TreeNode) {
    if node == nil {
        return
    }
    
    r.Ans = append(r.Ans, node.Val)
    r.dfs(node.Left)
    r.dfs(node.Right)
}